### ISSUE TITLE
MIRit

### ISSUE THREAD

**Issue by jacopo-ronchi** [Bioconductor/Contributions#3249]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/jacopo-ronchi/MIRit

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by lshep** [comment]
Please fix all ERROR, Warning or Notes in the above build report before a reviewer is assigned for an in depth review. thank you

**Comment by jacopo-ronchi** [comment]
> Please fix all ERROR, Warning or Notes in the above build report before a reviewer is assigned for an in depth review. thank you 

Yes sure, i am working on it. The error was caused because of missing IS_BIOC_BUILD_MACHINE variable in the SPB Renviron file. However, i described the issue on the Bioc-devel mailing list and, thanks to Lori, the environment variable has been added. Thus, the error won't occur during the next build.

Regarding the warning, i have cached different things and i am working to reduce R CMD Check on MacOS, which takes longer than 10 minutes. If you have any suggestion i would be glad to make the required changes.

Best regards,
Jacopo

**Comment by lshep** [comment]
ah sorry @jacopo-ronchi  I didn't connect the two with the request on the mailing list.  Cheers!

**Comment by PeteHaitch** [comment]
Hi @jacopo-ronchi,

Thank you for submitting **MIRit** to Bioconductor.

Overall, the package is in very good shape and close to being ready for acceptance.
Well done and thank you for making the effort, particularly with regard to the thorough documentation!

In my review below I have separated the issues into **Required** and **Recommended** points that I would ask you to address before the package can be accepted.
Would you please provide line-by-line comments to my initial review so that I know what changes I'm looking for in my re-review.

Cheers,
Pete

### Required

- [ ] More of the code chunks in the vignette need to be evaluated. I understand you've been battling to get package build/check under the time limits (which is very much appreciated!) but I identified a few chunks (see below) that can be safely evaluated without greatly increasing the run time[^1]. More importantly, when I ran the code myself I found that I got slightly different results to the 'pre-computed' results that are used in the vignette, which goes to highlight the important of using evaluated code chunks as much as possible to avoid the code and documentation falling out-of-sync.

````diff
diff --git a/vignettes/MIRit.Rmd b/vignettes/MIRit.Rmd
index 3f32903..1f223f9 100644
--- a/vignettes/MIRit.Rmd
+++ b/vignettes/MIRit.Rmd
@@ -246,7 +246,7 @@ Additionally, when we run these functions, we must define different arguments, n
 
 Following our example, let's calculate differentially expressed genes and differentially expressed miRNAs in thyroid cancer.
 
-```{r diffexp, eval=FALSE}
+```{r diffexp}
 ## perform differential expression for genes
 experiment <- performGeneDE(experiment,
                             group = "disease",
@@ -260,11 +260,6 @@ experiment <- performMirnaDE(experiment,
                              design = model)
 ```
 
-```{r, echo=FALSE}
-## load the example object
-experiment <- loadExamples()
-```
-
 If not specified, the `performMirnaDE()` and `performGeneDE()` functions will define differentially expressed genes/miRNAs as those having an adjusted p-value lower than 0.05, and an absolute log2 fold change higher than 1 (FC > 2). However, this behavior can be changed by tweaking the `pCutoff` parameter, that specifies the statistical significance threshold; the `logFC` parameter, which indicates the minimum log2 fold change that features must display for being considered as differentially expressed; and the `pAdjustment` parameter, which specifies the approach used for multiple testing correction (default is `fdr` to use the Benjamini-Hochberg method).
 
 ### Advanced parameters {#param}
@@ -522,7 +517,7 @@ Given the above, MIRit allows the prediction of miRNA-target interactions via th
 
 In our example, we are going to retrieve both predicted and validated interactions by using default settings.
 
-```{r targets, eval=FALSE}
+```{r targets, results='hide'}
 ## retrieve miRNA target genes
 experiment <- getTargets(experiment)
 ```
@@ -645,7 +640,7 @@ In MIRit, the `mirnaIntegration()` function automatically performs association t
 
 For example, we could use the Boschloo's exact to evaluate the inverse association between miRNA and gene expression values through a simple call to `mirnaIntegration()` function.
 
-```{r association, eval=FALSE}
+```{r association}
 ## perform a one-sided inverse association
 exp.association <- mirnaIntegration(experiment,
                                     test = "association",
@@ -661,7 +656,7 @@ Lastly, for unpaired data, the effect of DE-miRNAs on the expression of target g
 
 To perform the integrative analysis through rotation gene-set tests, we must simply set `test = "fry"` when calling `mirnaIntegration()` function.
 
-```{r fry, eval=FALSE}
+```{r fry}
 ## perform the integrative analysis through 'fry' method
 exp.fry <- mirnaIntegration(experiment,
                             test = "fry",
@@ -742,7 +737,7 @@ Before performing TAIPA, we need to create miRNA-augmented networks. To do so, M
 
 In our example, we want to use the significant miRNA-target pairs that we identified in Section \@ref(correlation) to augment biological pathways retrieved from the KEGG database.
 
-```{r augmented_networks, eval=FALSE}
+```{r augmented_networks}
 ## create miRNA-augmented networks using KEGG pathways
 networks <- preparePathways(experiment,
                             database = "KEGG",
````

[^1]: At least on my 2020 M1 MacBook Air, `R CMD build` only takes 1.5 minutes and `R CMD check` is < 3 minutes.

- [ ] I wonder if **MultiAssayExperiment** should be listed under `Depends` rather than under `Imports` because the *MirnaExperiment* class extends the *MultiAssayExperiment* class. Doing so would automatically provide the **MultiAssayExperiment** API when a user runs `library(MIRit)`. For example, when I was following along with the vignette, I was surprised I couldn't immediately run `experments()`, `assays()`, `colData()`, `sampleMap()`, etc. on the example *MirnaExperiment*.
- [ ] In the vignette, it states "MIRit uses the geneset package" but there's no package by that name and the link (https://bioconductor.org/packages/3.19/geneset) leads to a '404 page not found' error.

- [ ] Possibly related to the vignette using pre-computed results, but when I ran the  through the vignette and actually evalauted the code, the section demonstrating `findMirnaSNPs()` and `mirVariantPlot()` gave an error because there were no disease-related SNPs:

```r
association <- findMirnaSNPs(experiment, "Alzheimer disease")
Querying GWAS Catalog, this may take some time...
Finding genomic information of differentially expressed miRNAs...
No disease-related SNPs are present within DE-miRNA loci.

mirVariantPlot("rs2632516", association, showContext = TRUE)
Error: 'snpAssociation' must be a data.frame containing the list of SNPs occurring at DE-miRNA genes. To obtain this association you can use the 'findMirnaSNPs()' function. See ?findMirnaSNPs for details.
```

### Recommended

- [ ] Good work adding unit tests to your package. I recommend running `covr::report()` on the package code to identify untested/under-tested code.
- [ ] There are some warnings generated by code in the vignette, which are then included in the rendered version. I suggest taking a look at those to make sure its behaving as intended and if the warnings are unavoidable to perhaps add a comment to the vignette explaining why that is.
- [ ] The **limma**/**edgeR** authors explicitly recommend against applying a logFC cutoff to the results of a DE analysis[^2], so `performGeneDE(logFC = 1)`/`performMirnaDE(logFC = 1)` is perhaps a 'bad' default.

[^2]: ["A commonly used approach is to apply FDR and logFC cutoffs simultaneously. However this tends to favor lowly expressed genes, and also fails to control the FDR correctly."](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4934518/).
Instead, the authors developed `limma::treat()` and `edgeR::glmTreat()`; see their documentation for further details.
**DESeq2** may have similar functionality but I'm not as familiar with that package.

- [ ] Possibly related to the vignette using pre-computed results, the fry example produced no statistically significant association when I ran it; is that to be expected and as you intended?
- [ ] Minor: formatting of 'matrixarray' in the `show,MirnaExperiment-method` (see below) is perhaps not what you intended?

```r
> show(experiment)
An object of class MirnaExperiment, which extendsMultiAssayExperiment class and contains:

	- microRNA expression values: matrixarray with 371 rows and 16 columns
	- gene expression values: matrixarray with 2637 rows and 16 columns
	- samples metadata: DFrame with 16 rows and 5 columns
	- microRNA differential expression: data.frame with 371 rows and 5 columns
	- significant DE-miRNAs: character with 40 miRNA IDs
	- gene differential expression: data.frame with 2637 rows and 5 columns
	- significant genes: character with 267 gene IDs
	- microRNA targets: data.frame with 13605 rows and 6 columns
	- miRNA - gene integrative analysis: data.frame with 210 rows and 6 columns

MicroRNA and gene expression data derive from: paired samples
```

- [ ] Minor: Suggest clarifying in the documentation that **MIRit** uses **edgeR**'s quasi-likelihood framework (since **edgeR** also has the likelihood ratio test and exact test frameworks).
- [ ] Minor: In the vignette it says, "For example, the correlation analysis performed in Section 7.1.2 revealed how miR-146b-5p, the most upregulated miRNA, is inversely correlated with the expression of DIO2,", but I could not see this result ('miR-146b-5p' doesn't appear in that section as far as I can tell).
- [ ] Minor: For the `BugReports` field in the `DESCRIPTION`, some developers put the GitHub issues page for their package here, distinguishing 'bug reports' (i.e. issues with the package code and documentation, typically recommended to be posted to https://github.com/jacopo-ronchi/MIRit/issues) from 'user support' (typically recommended to be posted to https://support.bioconductor.org/tag/MIRit). But it's fine as it is and the choice is yours as a developer.

**Comment by jacopo-ronchi** [comment]
Dear @PeteHaitch,

First of all I would like to thank you for the detailed and competent review of my package. I really appreciated each of your suggestions as they were extremely useful to make MIRit better.

With this new update, all your concerns have been addressed, and below you will find detailed answers to each point.

---

> - [x] More of the code chunks in the vignette need to be evaluated. I understand you've been battling to get package build/check under the time limits (which is very much appreciated!) but I identified a few chunks (see below) that can be safely evaluated without greatly increasing the run time[1](https://github.com/Bioconductor/Contributions/issues/3249#user-content-fn-1-7cc8297eccfa24ba0754e3030b05810b). More importantly, when I ran the code myself I found that I got slightly different results to the 'pre-computed' results that are used in the vignette, which goes to highlight the important of using evaluated code chunks as much as possible to avoid the code and documentation falling out-of-sync.

Yes, you are correct. As you point out, there were slightly different results between the output of the vignette chunks and the precomputed objects. So the vignette has been completely redesigned and **every code chunk is now evaluated**. This way the vignette is completely dynamic and will not be out of sync with the documentation.

---

> - [x]  I wonder if MultiAssayExperiment should be listed under Depends rather than under Imports because the MirnaExperiment class extends the MultiAssayExperiment class. Doing so would automatically provide the MultiAssayExperiment API when a user runs library(MIRit). For example, when I was following along with the vignette, I was surprised I couldn't immediately run experments(), assays(), colData(), sampleMap(), etc. on the example MirnaExperiment.

I moved MultiAssayExperiment to Depends based on your advice.

---

> - [x] In the vignette, it states "MIRit uses the geneset package" but there's no package by that name and the link (https://bioconductor.org/packages/3.19/geneset) leads to a '404 page not found' error.

The geneset package is on CRAN, I fixed the link in the vignette.

---

> - [x] Possibly related to the vignette using pre-computed results, but when I ran the through the vignette and actually evalauted the code, the section demonstrating findMirnaSNPs() and mirVariantPlot() gave an error because there were no disease-related SNPs:

In the new vignette, the examples with SNPs are now evaluated and fully functional.

---

> - [x] Good work adding unit tests to your package. I recommend running covr::report() on the package code to identify untested/under-tested code.

I have done this. I would also like to increase the amount of tested code, but that would increase R CMD check times. I have posted a mail to the Bioc-devel mailing list about the strangely high check times on MacOS. So for now, I am waiting for updates on this issue before implementing new tests.

---

> - [x] There are some warnings generated by code in the vignette, which are then included in the rendered version. I suggest taking a look at those to make sure its behaving as intended and if the warnings are unavoidable to perhaps add a comment to the vignette explaining why that is.

This update fixes warnings. Now only one chunk throws a warning, which is expected and described in the vignette.

---

> - [x] The limma/edgeR authors explicitly recommend against applying a logFC cutoff to the results of a DE analysis[2](https://github.com/Bioconductor/Contributions/issues/3249#user-content-fn-2-7cc8297eccfa24ba0754e3030b05810b), so performGeneDE(logFC = 1)/performMirnaDE(logFC = 1) is perhaps a 'bad' default.

Fixed, now the default option is not to include logFC cutoff.

---

> - [x] Possibly related to the vignette using pre-computed results, the fry example produced no statistically significant association when I ran it; is that to be expected and as you intended?

Now this chunk is evaluated and returns significant results.

---

> - [x] Minor: formatting of 'matrixarray' in the show,MirnaExperiment-method (see below) is perhaps not what you intended?

Formatting is now fixed!

---

> - [x] Minor: Suggest clarifying in the documentation that MIRit uses edgeR's quasi-likelihood framework (since edgeR also has the likelihood ratio test and exact test frameworks).

You are absolutely correct. This information has been added to both the vignette and the documentation.

---

> - [x] Minor: In the vignette it says, "For example, the correlation analysis performed in Section 7.1.2 revealed how miR-146b-5p, the most upregulated miRNA, is inversely correlated with the expression of DIO2,", but I could not see this result ('miR-146b-5p' doesn't appear in that section as far as I can tell).

This was fixed in the new vignette.

---

> - [x] Minor: For the BugReports field in the DESCRIPTION, some developers put the GitHub issues page for their package here, distinguishing 'bug reports' (i.e. issues with the package code and documentation, typically recommended to be posted to https://github.com/jacopo-ronchi/MIRit/issues) from 'user support' (typically recommended to be posted to https://support.bioconductor.org/tag/MIRit). But it's fine as it is and the choice is yours as a developer.

Ok, no problem. I changed the link in the DESCRIPTION to receive bug reports to https://github.com/jacopo-ronchi/MIRit/issues.

---

In summary, everything has been addressed in version 0.99.13. The only suggestion I was not able to implement was to increase the unit tests. Currently, MIRit takes 4 minutes to check on the Linux builder, but is (weirdly) much slower on MacOS (10 minutes and 30 seconds). So I wrote an email to the Bioc-devel mailing list, and some members confirmed the lower performance on MacOS. So I will wait for their response before including a more extensive test suite.

Thanks again for your valuable feedback.
Jacopo

**Comment by PeteHaitch** [comment]
Thank you for making the requested changes, @jacopo-ronchi.
I'm happy to accept **MIRit** into Bioconductor.
Congratulations and thank you for your contribution!

Regarding the addition of unit tests increasing the check time, if you aren't aware then you might be interested to learn about 'long tests' (https://contributions.bioconductor.org/long-tests.html).

**Comment by lshep** [comment]
The default branch of your GitHub repository has been added to Bioconductor's
git repository as branch devel.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/jacopo-ronchi.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("MIRit")`. The package 'landing page' will be created at

https://bioconductor.org/packages/MIRit

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
